eth_abi/utils/grammar.py

```diff
@@ -0,0 +1,270 @@
+    grammar = grammar
+
+    def visit_tuple_type(self, node, visited_children):
+        # Ignore left and right parens
+        _, first, rest, _ = visited_children
+
+        return TupleType((first,) + rest, node=node)
+
+    def visit_next_type(self, node, visited_children):
+        # Ignore comma
+        _, abi_type = visited_children
+
+        return abi_type
+
+    def visit_basic_type(self, node, visited_children):
+        base, sub, arrlist = visited_children
+
+        return BasicType(base, sub, arrlist, node=node)
+
+    def visit_two_size(self, node, visited_children):
+        # Ignore "x"
+        first, _, second = visited_children
+
+        return first, second
+
+    def visit_const_arr(self, node, visited_children):
+        # Ignore left and right brackets
+        _, int_value, _ = visited_children
+
+        return (int_value,)
+
+    def visit_dynam_arr(self, node, visited_children):
+        return tuple()
+
+    def visit_alphas(self, node, visited_children):
+        return node.text
+
+    def visit_digits(self, node, visited_children):
+        return int(node.text)
+
+    def generic_visit(self, node, visited_children):
+        if isinstance(node.expr, expressions.OneOf):
+            # Unwrap value chosen from alternatives
+            return visited_children[0]
+
+        if isinstance(node.expr, expressions.Optional):
+            # Unwrap optional value or return `None`
+            if len(visited_children) != 0:
+                return visited_children[0]
+
+            return None
+
+        return tuple(visited_children)
+
+    @functools.lru_cache(maxsize=None)
+    def parse(self, type_str):
+        if not isinstance(type_str, str):
+            raise TypeError('Can only parse string values: got {}'.format(type(type_str)))
+
+        try:
+            return super().parse(type_str)
+        except parsimonious.ParseError as e:
+            raise ParseError(e.text, e.pos, e.expr)
+
+
+visitor = NodeVisitor()
+
+
+class ABIType:
+    __slots__ = ('node',)
+
+    def __init__(self, *, node=None):
+        self.node = node
+
+    def __repr__(self):
+        return '<{} {}>'.format(type(self).__qualname__, repr(str(self)))
+
+    def __eq__(self, other):
+        return (
+            type(self) is type(other) and
+            str(self) == str(other)
+        )
+
+    def __str__(self):
+        raise NotImplementedError('Must implement `__str__`')
+
+    def validate(self):
+        raise NotImplementedError('Must implement `validate`')
+
+    def invalidate(self, error_msg):
+        node = self.node
+
+        raise ABITypeError(
+            "For '{comp_str}' type at column {col} "
+            "in '{type_str}': {error_msg}".format(
+                comp_str=node.text,
+                col=node.start + 1,
+                type_str=node.full_text,
+                error_msg=error_msg,
+            ),
+        )
+
+
+class TupleType(ABIType):
+    __slots__ = ('components',)
```

Comment:
omg.  How did I not know about this!

http://book.pythontips.com/en/latest/__slots__magic.html

Looks like a 40-50% reduction in memory footprint!

Reply:
Yeah it seems like `__slots__` is ideal for small "immutable" container classes that only have a few attributes and many instances.  Constant time attribute lookups and space savings ftw.

Reply:
Although Python dicts in 3.6 share keys and are compact.  So I think it's less and less of a problem to just use normal classes with `__dict__` attrs.  Using `__slots__` is still a neat trick to have in your bag though.

Reply:
Also, it's looking like our `mypy` based type checking is going to be incompatible with `__slots__`.  See https://github.com/ethereum/py-evm/issues/520

Reply:
Is that still the consensus?  It seemed that @carver had found a way to make it work.

Reply:
Fine to leave in for now until we implement type hinting at which point we can see if we can use @carver 's solution.